pkg/kv/kvserver/batcheval/split_trigger_datadriven.go

```diff
@@ -0,0 +1,461 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package batcheval
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/abortspan"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/stateloader"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/datadriven"
+	"github.com/stretchr/testify/require"
+)
+
+/*
+TestSplitTriggerDatadriven tests the use of the split trigger. The test uses the
+following format:
+
+create-descriptor start=<key> end=<key> [replicas=[<int>,<int>,...]]
+----
+
+  Creates a range descriptor with the specified start and end keys and optional
+  replica list. The range ID is auto-assigned. If provided, replicas specify
+  NodeIDs for replicas of the range. Note that ReplicaIDs are assigned
+  incrementally starting from 1.
+
+create-split range-id=<int> split-key=<key>
+----
+
+  Creates a split for the specified range at the given split key, which entails
+  creating a SplitTrigger with both the LHS and RHS descriptors. Much like how
+  things work in CRDB, the LHS descriptor is created by narrowing the original
+  range and a new range descriptor is created for the RHS with the same replica
+  set.
+
+set-lease range-id=<int> replica=<int> [lease-type=leader-lease|epoch|expiration]
+----
+
+  Sets the lease for the specified range to the supplied replica. Note that the
+  replica parameter specifies NodeIDs, not to be confused with ReplicaIDs. By
+  default, the lease is of the leader-lease variety, but this may be overriden to
+  an epoch or expiration based lease by using the lease-type parameter. For now,
+  we treat the associated lease metadata as uninteresting.
+
+print-lease range-id=<int>
+----
+
+  Prints the leaseholder replica and lease type for the specified range.
+
+print-range-state
+----
+
+  Prints the entire range state of the test context.
+
+run-split-trigger range-id=<int>
+----
+  Executes the split trigger for the specified range. The output shows all
+  keys and values written to the batch as part of this.
+
+TODO(arul): For now, we do not change things like GC Threshold, GC Hint, etc.
```

Comment:
I feel like we don't need these combinatorics (as well as `set-lease` etc) in a datadriven test for testing RSE. Maybe there could be separate mini-tests for how each of these fields reacts to a split.

E.g. of the kind:

1. Generate a typical pre-split replica state (a shared/reusable func).
2. Fix it up (e.g. set the lease, GC hint, etc, differently).
3. Run split trigger.
4. Check the post-split state.
5. (optionally) Print the pre-split state and the output of the split trigger to `echotest`.